### PR TITLE
Handle more cases in dead member elim

### LINKED ISSUE #3282: Assertion in dead member elimination with spec constants
Hi,

I have this compute shader:
```
#version 450 core

layout(constant_id = 0) const int c_0 = 1;

layout(local_size_x_id = 1, local_size_y = 1, local_size_z = 1) in;

void main()
{
}
```

Compiling it with glslang without opts will give this:
```
              OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main"
               OpExecutionMode %main LocalSize 1 1 1
               OpSource GLSL 450
               OpName %main "main"
               OpName %c_0 "c_0"
               OpDecorate %c_0 SpecId 0
               OpDecorate %9 SpecId 1
               OpDecorate %gl_WorkGroupSize BuiltIn WorkgroupSize
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
        %int = OpTypeInt 32 1
        %c_0 = OpSpecConstant %int 1
       %uint = OpTypeInt 32 0
          %9 = OpSpecConstant %uint 1
     %uint_1 = OpConstant %uint 1
     %v3uint = OpTypeVector %uint 3
%gl_WorkGroupSize = OpSpecConstantComposite %v3uint %9 %uint_1 %uint_1
       %main = OpFunction %void None %3
          %5 = OpLabel
               OpReturn
               OpFunctionEnd
```

Using spirv-opt with -Os throws this assertion:
```
spirv-opt: /xxx/source/opt/eliminate_dead_members_pass.cpp:397: bool spvtools::opt::EliminateDeadMembersPass::UpdateConstantComposite(spvtools::opt::Instruction*): Assertion `inst->opcode() == SpvOpConstantComposite || inst->opcode() == SpvOpCompositeConstruct' failed.
```

Any idea what might be wrong?

### LINKED ISSUE #3284: Unimplemented path in dead member elimination pass (spec constants)
Hi,

I'm hitting an assertion when using -Os in spirv-opt:
```
spirv-opt: /xxx/source/opt/eliminate_dead_members_pass.cpp:298: spvtools::opt::EliminateDeadMembersPass::RemoveDeadMembers()::<lambda(spvtools::opt::Instruction*)>: Assertion `false && "Not yet implemented."' failed.
```

The shader is quite simple. It was compiled with glslang without opts:
```
#version 450 core

layout(constant_id = 0) const int c_0 = 1;
layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;

layout(set = 0, binding = 2, std140) buffer ss0_
{
	float u_s[c_0];
};

void main()
{
	u_s[gl_LocalInvocationID.x] = gl_LocalInvocationID.y + c_0;
}
```

SPIRV for it:
```
               OpCapability Shader
          %1 = OpExtInstImport "GLSL.std.450"
               OpMemoryModel Logical GLSL450
               OpEntryPoint GLCompute %main "main" %gl_LocalInvocationID
               OpExecutionMode %main LocalSize 1 1 1
               OpSource GLSL 450
               OpName %main "main"
               OpName %c_0 "c_0"
               OpName %ss0_ "ss0_"
               OpMemberName %ss0_ 0 "u_s"
               OpName %_ ""
               OpName %gl_LocalInvocationID "gl_LocalInvocationID"
               OpName %c_0 "c"
               OpDecorate %c_0 SpecId 0
               OpDecorate %_arr_float_c_0 ArrayStride 16
               OpMemberDecorate %ss0_ 0 Offset 0
               OpDecorate %ss0_ BufferBlock
               OpDecorate %_ DescriptorSet 0
               OpDecorate %_ Binding 2
               OpDecorate %gl_LocalInvocationID BuiltIn LocalInvocationId
               OpDecorate %gl_WorkGroupSize BuiltIn WorkgroupSize
       %void = OpTypeVoid
          %3 = OpTypeFunction %void
      %float = OpTypeFloat 32
        %int = OpTypeInt 32 1
        %c_0 = OpSpecConstant %int 1
%_arr_float_c_0 = OpTypeArray %float %c_0
       %ss0_ = OpTypeStruct %_arr_float_c_0
%_ptr_Uniform_ss0_ = OpTypePointer Uniform %ss0_
          %_ = OpVariable %_ptr_Uniform_ss0_ Uniform
      %int_0 = OpConstant %int 0
       %uint = OpTypeInt 32 0
     %v3uint = OpTypeVector %uint 3
%_ptr_Input_v3uint = OpTypePointer Input %v3uint
%gl_LocalInvocationID = OpVariable %_ptr_Input_v3uint Input
     %uint_0 = OpConstant %uint 0
%_ptr_Input_uint = OpTypePointer Input %uint
     %uint_1 = OpConstant %uint 1
         %25 = OpSpecConstantOp %uint IAdd %c_0 %uint_0
%_ptr_Uniform_float = OpTypePointer Uniform %float
%gl_WorkGroupSize = OpConstantComposite %v3uint %uint_1 %uint_1 %uint_1
       %main = OpFunction %void None %3
          %5 = OpLabel
         %20 = OpAccessChain %_ptr_Input_uint %gl_LocalInvocationID %uint_0
         %21 = OpLoad %uint %20
         %23 = OpAccessChain %_ptr_Input_uint %gl_LocalInvocationID %uint_1
         %24 = OpLoad %uint %23
         %26 = OpIAdd %uint %24 %25
         %27 = OpConvertUToF %float %26
         %29 = OpAccessChain %_ptr_Uniform_float %_ %int_0 %21
               OpStore %29 %27
               OpReturn
               OpFunctionEnd
```

Any plans on implementing this (@s-perron ?)? I'm afraid specialization constants have many issues in the optimizer (Another issue is #3282).

### PR DESCRIPTION
**Pull request by s-perron**
- Rewrite composite insert and extract operations on SpecConstnatOp.
- Leaves assert for Access chain instructions, which are only allowed
for kernels.
- Other operations do not require any extra code will no longer cause an
assert.

Fixes #3284.
Fixes #3282.

### CODE REVIEW

File: source/opt/eliminate_dead_members_pass.cpp [hunk 1]
```diff
@@ -153,13 +169,21 @@ void EliminateDeadMembersPass::MarkMembersAsLiveForCopyMemory(
...
-  assert(inst->opcode() == SpvOpCompositeExtract);
+  assert(inst->opcode() == SpvOpCompositeExtract ||
+      (inst->opcode() == SpvOpSpecConstantOp &&
+          inst->GetSingleWordInOperand(kSpecConstOpOpcodeIdx) ==
+              SpvOpCompositeExtract));
+
+  uint32_t first_operand = 0;
+  if (inst->opcode() == SpvOpSpecConstantOp) {
+    first_operand = 1;
+  }
```
Comment (kuhar) [inline source/opt/eliminate_dead_members_pass.cpp:180]: nit: maybe
```
const uint32_t first_operand = (inst->opcode() == SpvOpSpecConstantOp) ? 1 : 0;
```

File: source/opt/eliminate_dead_members_pass.cpp [hunk 2]
```diff
@@ -295,7 +319,23 @@ bool EliminateDeadMembersPass::RemoveDeadMembers() {
...
+          case SpvOpAccessChain:
+          case SpvOpInBoundsAccessChain:
+          case SpvOpPtrAccessChain:
+          case SpvOpInBoundsPtrAccessChain:
+            assert(false && "Not implemented yet.");
+            break;
+          default:
+            break;
+        }
+        // assert(false && "Not yet implemented.");
```
Comment (kuhar) [inline source/opt/eliminate_dead_members_pass.cpp:338]: Why is this assertion commented out?

File: source/opt/eliminate_dead_members_pass.cpp (hunk 1 above)
Comment (s-perron) [reply source/opt/eliminate_dead_members_pass.cpp:180]: done.

File: source/opt/eliminate_dead_members_pass.cpp (hunk 2 above)
Comment (s-perron) [reply source/opt/eliminate_dead_members_pass.cpp]: This an the following lines are a relic from before that should be deleted.

**Pull request merged**